### ISSUE TITLE
Calculated remaining charge time incorrect

### ISSUE THREAD

**Issue by mibo-wb-gh** [bimmerconnected/bimmer_connected#353]
### Describe the issue

Calculated remaining charge time incorrect

Shown remaining calculated charge time in HA: 10.97h

100% Charged shown on My BMW App: 05:44 AM

So remaining charge time in HA should be: 8 hours and 37 minutes

Fingerprint created with version 0.8.0.0b10 while charging is ongoing

Local time at the moment of the fingerprint: 21:07

Timezone CET - EUROPE/BERLIN

330e xDrive Touring (G21) 2020
rest of world

[vehicles_v2_bmw_0.txt](https://github.com/bimmerconnected/bimmer_connected/files/7565748/vehicles_v2_bmw_0.txt)

![20211118_211440](https://user-images.githubusercontent.com/87424844/142492044-395e44d6-dcc6-4485-8e20-8bb31247f804.jpg)

Maybe the error could come from the following in the fingerprint in row 453 the following is stated:
"infoLabel": "100% at ~08:23 AM",

-> 100% should be listed as 05:44 AM

### Expected behavior

So remaining charge time in HA should be: 8 hours and 37 minutes

### Which Home Assistant version are you using?

core-2021.11.4

### What is your region?

Rest of world

### ConnectedDrive website

- [X] I can still successfully login to the BMW ConnectedDrive website and the car status is available there.
- [X] I have ConnectedDrive enabled for my vehicle.

### Number of cars

- [ ] I have 2 or more cars linked to the ConnectedDrive account.
- [ ] I have a Mini vehicle linked to my account.

### Output of bimmer_connected fingerprint

```txt
[
  {
    "a4aType": "NOT_SUPPORTED",
    "bodyType": "G21",
    "brand": "BMW",
    "capabilities": {
      "canRemoteHistoryBeDeleted": false,
      "climateNow": {
        "executionMessage": "Do you want to ventilate now? Remote functions may take a few seconds.",
        "executionPopup": {
          "executionMessage": "Turn pre-conditioning on now? Remote functions may take a few seconds.",
          "iconId": 59733,
          "popupType": "DIALOG",
          "primaryButtonText": "Start",
          "secondaryButtonText": "Cancel",
          "title": "Start Climatization"
        },
        "executionStopPopup": {
          "executionMessage": "Stop climate control in your vehicle now? Remote functions may take a few seconds.",
          "title": "Climate control is running"
        },
        "isEnabled": true,
        "isPinAuthenticationRequired": false
      },
      "horn": {
        "executionMessage": "Using your horn is only allowed in certain situations in many countries. Responsibility for the use and adherence to the respective regulations lies solely with you as the user. \n\nDo you want to use the horn now? Remote functions may take a few seconds.",
        "isEnabled": true,
        "isPinAuthenticationRequired": false
      },
      "isBmwChargingSupported": true,
      "isCarSharingSupported": false,
      "isChargeNowForBusinessSupported": true,
      "isChargingHistorySupported": true,
      "isChargingHospitalityEnabled": false,
      "isChargingLoudnessEnable": false,
      "isChargingPlanSupported": true,
      "isChargingPowerLimitEnable": false,
      "isChargingSettingsEnabled": false,
      "isChargingTargetSocEnable": false,
      "isCustomerEsimSupported": false,
      "isDCSContractManagementSupported": true,
      "isDataPrivacyEnabled": false,
      "isEasyChargeSupported": false,
      "isEvGoChargingSupported": false,
      "isFindChargingEnabled": true,
      "isMiniChargingSupported": false,
      "isRemoteHistorySupported": true,
      "isRemoteServicesActivationRequired": false,
      "isRemoteServicesBookingRequired": false,
      "isScanAndChargeSupported": true,
      "lastStateCall": {
        "isNonLscFeatureEnabled": false,
        "lscState": "ACTIVATED"
      },
      "lights": {
        "executionMessage": "Flash headlights now? Remote functions may take a few seconds.",
        "isEnabled": true,
        "isPinAuthenticationRequired": false
      },
      "lock": {
        "executionMessage": "Lock your vehicle now? Remote functions may take a few seconds.",
        "isEnabled": true,
        "isPinAuthenticationRequired": false
      },
      "remote360": {
        "isComingSoonEnabled": false,
        "isDataPrivacyEnabled": false,
        "isEnabled": true,
        "isPinAuthenticationRequired": false,
        "isToggleEnabled": true
      },
      "remoteSoftwareUpgrade": {
        "isEnabled": true,
        "isPinAuthenticationRequired": false
      },
      "sendPoi": {
        "executionMessage": "Send POI now? Remote functions may take a few seconds.",
        "isEnabled": true,
        "isPinAuthenticationRequired": false
      },
      "speechThirdPartyAlexa": {
        "executionMessage": "Activate Alexa now? Remote functions may take a few seconds.",
        "isEnabled": true,
        "isPinAuthenticationRequired": false
      },
      "unlock": {
        "executionMessage": "Unlock your vehicle now? Remote functions may take a few seconds.",
        "isEnabled": true,
        "isPinAuthenticationRequired": true
      },
      "vehicleFinder": {
        "executionMessage": "Find your vehicle now? Remote functions may take a few seconds.",
        "isEnabled": true,
        "isPinAuthenticationRequired": false
      }
    },
    "connectedDriveServices": [
      "WIFI_HOTSPOT_SERVICE"
    ],
    "driveTrain": "PLUGIN_HYBRID",
    "driverGuideInfo": {
      "androidAppScheme": "com.bmwgroup.driversguide.row",
      "androidStoreUrl": "https://play.google.com/store/apps/details?id=com.bmwgroup.driversguide.row",
      "iosAppScheme": "bmwdriversguide:///open",
      "iosStoreUrl": "https://apps.apple.com/de/app/id714042749?mt=8",
      "title": "BMW\nDriver's Guide"
    },
    "exFactoryILevel": "S18A-20-07-548",
    "exFactoryPUStep": "0720",
    "headUnit": "MGU",
    "hmiVersion": "id7",
    "iStep": "S18A-21-07-550",
    "isLscSupported": true,
    "isMappingPending": false,
    "isMappingUnconfirmed": false,
    "model": "330e xDrive",
    "properties": {
      "areDoorsClosed": true,
      "areDoorsLocked": true,
      "areDoorsOpen": false,
      "areWindowsClosed": true,
      "chargingState": {
        "chargePercentage": 29,
        "isChargerConnected": true,
        "state": "CHARGING",
        "type": "NOT_AVAILABLE"
      },
      "checkControlMessages": [],
      "climateControl": {
        "activity": "INACTIVE"
      },
      "combinedRange": {
        "distance": {
          "units": "KILOMETERS",
          "value": 257
        }
      },
      "combustionRange": {
        "distance": {
          "units": "KILOMETERS",
          "value": 257
        }
      },
      "doorsAndWindows": {
        "doors": {
          "driverFront": "CLOSED",
          "driverRear": "CLOSED",
          "passengerFront": "CLOSED",
          "passengerRear": "CLOSED"
        },
        "hood": "CLOSED",
        "moonroof": "CLOSED",
        "trunk": "CLOSED",
        "windows": {
          "driverFront": "CLOSED",
          "driverRear": "CLOSED",
          "passengerFront": "CLOSED",
          "passengerRear": "CLOSED"
        }
      },
      "electricRange": {
        "distance": {
          "units": "KILOMETERS",
          "value": 7
        }
      },
      "electricRangeAndStatus": {
        "chargePercentage": 29,
        "distance": {
          "units": "KILOMETERS",
          "value": 7
        }
      },
      "fuelLevel": {
        "units": "LITERS",
        "value": 23
      },
      "fuelPercentage": {
        "value": 66
      },
      "inMotion": false,
      "isServiceRequired": false,
      "lastUpdatedAt": "2021-11-18T20:07:59Z",
      "originCountryISO": "DE",
      "serviceRequired": [
        {
          "dateTime": "2022-10-01T00:00:00.000Z",
          "distance": {
            "units": "KILOMETERS",
            "value": 6000
          },
          "status": "OK",
          "type": "OIL"
        },
        {
          "dateTime": "2024-10-01T00:00:00.000Z",
          "distance": {
            "units": "KILOMETERS",
            "value": 39000
          },
          "status": "OK",
          "type": "VEHICLE_CHECK"
        },
        {
          "status": "OK",
          "type": "TIRE_WEAR_FRONT"
        },
        {
          "status": "OK",
          "type": "TIRE_WEAR_REAR"
        },
        {
          "dateTime": "2023-10-01T00:00:00.000Z",
          "status": "OK",
          "type": "BRAKE_FLUID"
        }
      ],
      "tires": {
        "frontLeft": {
          "details": {
            "dimension": "225/45 R18 95V XL",
            "manufacturer": "Pirelli",
            "manufacturingWeek": "35 / 19",
            "maxSpeed": "240 km/h",
            "mountingDate": "11/16/2021",
            "optimizedForOemBmw": "Yes",
            "partNumber": "2461777",
            "season": "Winter tires",
            "treadDesign": "SOTTOZERO 3"
          },
          "status": {
            "currentPressure": 250,
            "localizedCurrentPressure": "2.5 bar",
            "localizedTargetPressure": "2.7 bar",
            "targetPressure": 270,
            "wear": 0
          }
        },
        "frontRight": {
          "details": {
            "dimension": "225/45 R18 95V XL",
            "manufacturer": "Pirelli",
            "manufacturingWeek": "27 / 19",
            "maxSpeed": "240 km/h",
            "mountingDate": "11/16/2021",
            "optimizedForOemBmw": "Yes",
            "partNumber": "2461777",
            "season": "Winter tires",
            "treadDesign": "SOTTOZERO 3"
          },
          "status": {
            "currentPressure": 240,
            "localizedCurrentPressure": "2.4 bar",
            "localizedTargetPressure": "2.7 bar",
            "targetPressure": 270,
            "wear": 0
          }
        },
        "rearLeft": {
          "details": {
            "dimension": "255/40 R18 99V XL",
            "manufacturer": "Pirelli",
            "manufacturingWeek": "17 / 19",
            "maxSpeed": "240 km/h",
            "mountingDate": "11/16/2021",
            "optimizedForOemBmw": "Yes",
            "partNumber": "2461778",
            "season": "Winter tires",
            "treadDesign": "SOTTOZERO 3"
          },
          "status": {
            "currentPressure": 290,
            "localizedCurrentPressure": "2.9 bar",
            "localizedTargetPressure": "2.9 bar",
            "targetPressure": 290,
            "wear": 0
          }
        },
        "rearRight": {
          "details": {
            "dimension": "255/40 R18 99V XL",
            "manufacturer": "Pirelli",
            "manufacturingWeek": "26 / 19",
            "maxSpeed": "240 km/h",
            "mountingDate": "11/16/2021",
            "optimizedForOemBmw": "Yes",
            "partNumber": "2461778",
            "season": "Winter tires",
            "treadDesign": "SOTTOZERO 3"
          },
          "status": {
            "currentPressure": 240,
            "localizedCurrentPressure": "2.4 bar",
            "localizedTargetPressure": "2.9 bar",
            "targetPressure": 290,
            "wear": 0
          }
        }
      },
      "vehicleLocation": {
        "address": {
          "formatted": "some_formatted_address"
        },
        "coordinates": {
          "latitude": 12.3456,
          "longitude": 34.5678
        },
        "heading": 123
      }
    },
    "puStep": "0721",
    "status": {
      "chargingProfile": {
        "chargingControlType": "weeklyPlanner",
        "chargingMode": "immediateCharging",
        "chargingPreference": "noPreSelection",
        "chargingSettings": {
          "hospitality": "NO_ACTION",
          "idcc": "NO_ACTION",
          "isAcCurrentLimitActive": false,
          "targetSoc": 100
        },
        "climatisationOn": false,
        "departureTimes": [
          {
            "action": "deactivate",
            "id": 1,
            "timeStamp": {
              "hour": 0,
              "minute": 0
            },
            "timerWeekDays": []
          },
          {
            "action": "deactivate",
            "id": 2,
            "timeStamp": {
              "hour": 0,
              "minute": 0
            },
            "timerWeekDays": []
          },
          {
            "action": "deactivate",
            "id": 3,
            "timeStamp": {
              "hour": 0,
              "minute": 0
            },
            "timerWeekDays": []
          },
          {
            "action": "deactivate",
            "id": 4,
            "timerWeekDays": []
          }
        ],
        "reductionOfChargeCurrent": {
          "end": {
            "hour": 0,
            "minute": 0
          },
          "start": {
            "hour": 0,
            "minute": 0
          }
        }
      },
      "checkControlMessages": [
        {
          "criticalness": "nonCritical",
          "iconId": 60117,
          "state": "OK",
          "title": "Tires"
        },
        {
          "criticalness": "nonCritical",
          "iconId": 60197,
          "state": "OK",
          "title": "Engine Oil"
        }
      ],
      "checkControlMessagesGeneralState": "No Issues",
      "currentMileage": {
        "formattedMileage": "27351",
        "mileage": 27351,
        "units": "km"
      },
      "doorsAndWindows": [
        {
          "criticalness": "nonCritical",
          "iconId": 59757,
          "state": "Locked",
          "title": "Lock status"
        },
        {
          "criticalness": "nonCritical",
          "iconId": 59722,
          "state": "Closed",
          "title": "All doors"
        },
        {
          "criticalness": "nonCritical",
          "iconId": 59725,
          "state": "Closed",
          "title": "All windows"
        },
        {
          "criticalness": "nonCritical",
          "iconId": 59706,
          "state": "Closed",
          "title": "Hood"
        },
        {
          "criticalness": "nonCritical",
          "iconId": 59704,
          "state": "Closed",
          "title": "Trunk"
        },
        {
          "criticalness": "nonCritical",
          "iconId": 59705,
          "state": "Closed",
          "title": "Sunroof"
        }
      ],
      "doorsGeneralState": "Locked",
      "fuelIndicators": [
        {
          "chargingType": null,
          "iconOpacity": "high",
          "infoIconId": 59691,
          "infoLabel": "Combined Range",
          "isCircleIcon": false,
          "isInaccurate": false,
          "levelIconId": null,
          "levelUnits": null,
          "levelValue": null,
          "mainBarValue": 0,
          "rangeIconId": 59691,
          "rangeUnits": "km",
          "rangeValue": "257",
          "secondaryBarValue": 0,
          "showsBar": false
        },
        {
          "barType": null,
          "chargingStatusIndicatorType": "CHARGING",
          "chargingStatusType": "CHARGING",
          "chargingType": "charging",
          "iconOpacity": "high",
          "infoIconId": 59689,
          "infoLabel": "100% at ~08:23 AM",
          "isCircleIcon": true,
          "isInaccurate": true,
          "levelIconId": 59689,
          "levelUnits": "%",
          "levelValue": "29",
          "mainBarValue": 29,
          "rangeIconId": 59683,
          "rangeUnits": "km",
          "rangeValue": "7",
          "secondaryBarValue": 0,
          "showBarGoal": false,
          "showsBar": true
        },
        {
          "chargingType": null,
          "iconOpacity": "high",
          "infoIconId": 59930,
          "infoLabel": "Fuel Level",
          "isCircleIcon": false,
          "isInaccurate": false,
          "levelIconId": 59682,
          "levelUnits": "%",
          "levelValue": "66",
          "mainBarValue": 66,
          "rangeIconId": 59681,
          "rangeUnits": "km",
          "rangeValue": "250",
          "secondaryBarValue": 0,
          "showsBar": true
        }
      ],
      "issues": {},
      "lastUpdatedAt": "2021-11-18T20:07:59Z",
      "recallExternalUrl": null,
      "recallMessages": [],
      "requiredServices": [
        {
          "criticalness": "nonCritical",
          "iconId": 60197,
          "id": "Oil",
          "longDescription": "Next service due after the specified distance or date.",
          "subtitle": "some_road \u2022 duration \u2022 -- EUR",
          "title": "Engine oil"
        },
        {
          "criticalness": "nonCritical",
          "iconId": 60215,
          "id": "VehicleCheck",
          "longDescription": "Next vehicle check due after the specified distance or date.",
          "subtitle": "some_road \u2022 duration \u2022 -- EUR",
          "title": "Vehicle check"
        },
        {
          "criticalness": "nonCritical",
          "iconId": 60447,
          "id": "TireWearFront",
          "subtitle": "some_road \u2022 duration \u2022 -- EUR",
          "title": "Tire service, front tires"
        },
        {
          "criticalness": "nonCritical",
          "iconId": 60447,
          "id": "TireWearRear",
          "subtitle": "some_road \u2022 duration \u2022 -- EUR",
          "title": "Tire service, rear tires"
        },
        {
          "criticalness": "nonCritical",
          "iconId": 60223,
          "id": "BrakeFluid",
          "longDescription": "Next service due by the specified date.",
          "subtitle": "some_road \u2022 duration \u2022 -- EUR",
          "title": "Brake fluid"
        }
      ],
      "timestampMessage": "Updated from vehicle 11/18/2021 09:07 PM"
    },
    "telematicsUnit": "ATM02",
    "themeSpecs": {
      "vehicleStatusBackgroundColor": {
        "blue": 201,
        "green": 94,
        "red": 40
      }
    },
    "vin": "some_vin",
    "year": 2020
  }
]
```

**Comment by rikroe** [comment]
Not really sure if we can do anything about it - except there is another endpoint that provides a better result.

I thought it could have something to do with wrong time zones, but that would only be the case if the minutes would be the same...

**Comment by mibo-wb-gh** [comment]
Looks like BMW stopped supplying the remaining chargetime. Odd that it is display correctly in the My BMW App

How are you calculation the remaining chargetime at the moment ?

**Comment by nottledim** [comment]
`status.charging_time_remaining` is there.  I don't know if it's populated or what units are.  I compute this myself from percentage battery (`charging_level_hv`).  BWM don't seem to be able to compute it correctly for actual charge rate anyway! 

```
minutes to go =>  (100 - <percent>) * <battery capacity (kWh)> * 60 / <charge rate (kW)> / 100
 ```

**Comment by mibo-wb-gh** [comment]
I can't find a "status.charging_time_remaining" entity. 

Of course there could be a difference between the output of a BEV and PHEV

The following entity is display with me:

"sensor.330e_xdrive_charging_time_remaining"

Which showed the following yesterday evening: 10.97h

It should have shown 8 hours 37 minutes remaining

At the moment it shows: 0.00h as the car is not connected to a charging point

I will try your suggestion to calculate it myself. Thanks !

**Comment by rikroe** [comment]
Calculating by yourself only works if you now the actual charging rate. Unfortunately, the API gives no indication on how fast a car is charging.

**Comment by nottledim** [comment]
True. You do know what sort of charger you're connected to which the API doesn't. Slow, fast, rapid, 3 phase etc.  In ha you can use an input select drop-down to set the nominal rate and then compute the duration and end time from the battery level percentage.  It works well except right at the end of the charge when the charge rate falls off.

**Comment by rikroe** [comment]
That's correct.

I just was able to double check the information the app is showing vs. the API: at least on the vehicle (an i3) I was able to test with, the value displayed in the app & the value in `fuelIndicators` were the same.

The only thing that I could trying to debug the connctions made for a newer PHEV, but that requires either giving me the credentials (rest of world) or sharing the vehicle with me (us), at least for a short time.

**Comment by mibo-wb-gh** [comment]
330e xDrive Touring (G21) PHEV / Plugin Hybrid Production Summer 2020 / Battery capacity: 12kWh

Charging at home (230V) : restricted to 6A which results in 1,4 KW/h
Charging at home (230V): allowing 10A which results in 2,3 KW/h
Charging at a chargestation: always using the maximum16A which results in 3,6KW/h

Maybe these stats help, otherwise we could take about getting access to the BMW Data via my login.

**Comment by rikroe** [comment]
Just fixed AM/PM parsing of the charging end times, please try again with the newest custom component/library.

**Comment by rikroe** [comment]
Please check the new sensor `charging_end_time` in the latest custom component. If you want to see the value returned by the BMW API, please also activate the `charging_end_time_mybmw` sensor from device settings in HA.

**Comment by rikroe** [comment]
Please check with the latest custom component. You should be able to activate `charging_end_time_original` from the entities list (hidden by default).
https://github.com/bimmerconnected/ha_custom_component/releases/tag/20211207.1

**Comment by rikroe** [comment]
@mibo-wb-gh I would like to come back to your offer regarding your BMW login. Can you shoot me an email to rikro@gmx.net or contact me on Telegram (@rikro2308)?

**Comment by rikroe** [comment]
Please check again with the latest custom component. To display `charging_end_time` correctly, you need to use Home Assistant **>=2021.12.0**.

**Comment by mibo-wb-gh** [comment]
Works perfect with HA 2021.12.0 and updated HACS BMW Connected Drive BETA !

Thanks a million for all the work everybody has put in to this !!!

**Comment by rikroe** [comment]
This issue should be fixed with HA 2021.12.4. Please uninstall the custom component if you are still using it.
